### PR TITLE
Trim managedFields in controller-manager

### LINKED ISSUE #118454: kube-controller-manager should not cache managedFields
### What would you like to be added?

Remove [`managedFields`](https://github.com/kubernetes/apimachinery/blob/v0.27.2/pkg/apis/meta/v1/types.go#L271) from the cached used by controllers in `kube-controller-manager`. If feasible, this work will be applied to other components having similar use patterns but have no use of `managedFields` at all.

### Why is this needed?

Roughly the content in `managedFields` is a copy of whole object so it takes roughly 40% of total object size, which is a significant memory overhead. Removing it for controllers can improve memory efficiency

### PR DESCRIPTION
**Pull request by linxiulei**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Improve memory efficiency for `kube-controller-manager`

#### Which issue(s) this PR fixes:

Fixes #118454

#### Does this PR introduce a user-facing change?

```release-note
Improve memory usage of kube-controller-manager by dropping the `.metadata.managedFields` field that kube-controller-manager doesn't require.
```

### CODE REVIEW

File: staging/src/k8s.io/controller-manager/pkg/informerfactory/informer_factory.go [hunk 1]
```diff
@@ -42,6 +43,18 @@ func (i *informerFactory) ForResource(resource schema.GroupVersionResource) (inf
 }
 
 func (i *informerFactory) Start(stopCh <-chan struct{}) {
+	trimFunc := func(obj interface{}) (interface{}, error) {
```
Comment (linxiulei) [inline staging/src/k8s.io/controller-manager/pkg/informerfactory/informer_factory.go:46]: @apelisse this code removes ManagedFields, PTAL.

Comment (apelisse) [reply staging/src/k8s.io/controller-manager/pkg/informerfactory/informer_factory.go:46]: Ah, yes, that's what I missed, thanks :-)

File: staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go [hunk 2]
```diff
@@ -257,6 +257,17 @@ func (f *sharedInformerFactory) InformerFor(obj {{.runtimeObject|raw}}, newFunc
   return informer
 }
 
+// ExistingInformers returns all existing shared informers.
+func (f *sharedInformerFactory) ExistingInformers() map[{{.schemaGroupVersionResource|raw}}]GenericInformer {
```
Comment (alexzielenski) [inline staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go:261]: Rather than add `ExistingInformers` to the factory, I personally would have added a new `SharedInformerOption` that allows the user to provide a closure like `func(informers.GenericInformer)` to run against the informer when the informer is created by the factory in `InformerFor`

**Comment by czybjtu** [comment]
should we modify every informer according on your way which we don't use managedFileld?

File: staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go (hunk 2 above)
Comment (linxiulei) [reply staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go]: Great suggestion. I added 'WithTransform' SharedInformerOption to make this option more precise.

**Comment by linxiulei** [comment]
> should we modify every informer according on your way which we don't use managedFileld?

Yes, that'd be my plan

**Comment by linxiulei** [comment]
Cleaned up and ready to review. I'd appreciate a "ok-to-test", thanks in advance.

File: staging/src/k8s.io/client-go/metadata/metadatainformer/informer.go [hunk 3]
```diff
@@ -31,6 +32,33 @@ import (
...
+	}
+}
+
+// WithNamespace limits the metadataSharedInformerFactory to the specified namespace.
+func WithNamespace(namespace string) SharedInformerOption {
+	return func(factory *metadataSharedInformerFactory) *metadataSharedInformerFactory {
+		factory.namespace = namespace
+		return factory
+	}
+}
```
Comment (apelisse) [inline staging/src/k8s.io/client-go/metadata/metadatainformer/informer.go:52]: How is this related?

Comment (linxiulei) [reply staging/src/k8s.io/client-go/metadata/metadatainformer/informer.go:52]: No, this is not related. But I'd like to make metadataSharedInformerFactory have identical SharedInformerOption(s) with SharedInformerFactory for consistency.

**Comment by apelisse** [comment]
Can we make them in separate PRs then? I'd rather do one thing at a time. Happy to lgtm/approve otherwise.

**Comment by linxiulei** [comment]
> Can we make them in separate PRs then? I'd rather do one thing at a time. Happy to lgtm/approve otherwise.

Sure, I removed unnecessary changes and rebased. Thanks!

**Comment by apelisse** [comment]
thanks
/lgtm
/approve

**Comment by alculquicondor** [comment]
/hold

For the implication of stripping the managedFields when
- KCM controllers use Update
- Out-of-tree controllers use SSA.

**Comment by alculquicondor** [comment]
FYI,
I'm supporting this change for kube-scheduler https://github.com/kubernetes/kubernetes/pull/119556

However, the change in kcm needs approval from SIG Architecture, IIUC.

**Comment by alculquicondor** [comment]
Ah, @wojtek-t's approval should suffice
/hold cancel

**Comment by aojea** [comment]
This seems better to wait for next release in 1.29, seems like a change risky for this part of the release cycle

/cc @liggitt

**Comment by alculquicondor** [comment]
Agree. As it stands, this is currently blocked on the code freeze, as intended.

**Comment by liggitt** [comment]
A data race on resync was just reported in use of WithTransform in https://github.com/kubernetes/client-go/issues/1348 that may affect this use

**Comment by wojtek-t** [comment]
heh...
the problem is that for Resync operation we take object that is already existing in the store (DeltaFifo in this case).
And there is chance of touching (reading) that object at the same time.

We effectively can't call Transform for SYNC events here:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/delta_fifo.go#L456

**Comment by liggitt** [comment]
> We effectively can't call Transform for SYNC events here:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/delta_fifo.go#L456

Or delete events where the object is DeletedTombstoneFinalStateUnknown, and maybe also for Replace (the docs there are unclear)

**Comment by nirvana274** [comment]
> > Or delete events where the object is DeletedFinalStateUnknown, and maybe also for Replace (the docs there are unclear - "In cases involving Replace(), such an object can come through multiple times.")
> 
> Good point on the DeletedFinalStateUnknown.
> 
> Regarding Replace() - this is probably true in theory, but in practice we only replace with a net new result of List call, so in practice it's not a problem [although we should clarify the contract there or sth]
